File: packages/silo-client/test/media/media-examples.test.ts
import { describe, expect, test } from "bun:test";
import { MediaInUseError } from "../../src/errors/media-in-use-error";
import { Media } from "../../src/media/media";
import { MediaReference } from "../../src/media/media-reference";
import { Transport } from "../../src/transport/transport";
import { StubFetch } from "../support/stub-fetch";
import { StubResponse } from "../support/stub-response";

/**
 * Every media example in the README, compiled and run against a stub
 * transport. Drives `Media` directly, which is what `silo.media` is.
 */
const assetPayload = (overrides: Record<string, unknown> = {}) => ({
  id: "01ASSET",
  filename: "hero.png",
  folder: "heroes",
  blob_key: "01ASSET.hero.png",
  size: 2048,
  content_type: "image/png",
  hash: "abc123",
  state: "active",
  tags: ["banner"],
  url: "http://localhost:8090/media/01ASSET",
  created_at: "2026-01-01T00:00:00.000Z",
  updated_at: "2026-01-01T00:00:00.000Z",
  usage_count: 1,
  ...overrides,
});

test("the plan's Media section, end to end", async () => {
  const stubFetch = new StubFetch();
  const transport = new Transport({ url: "http://localhost:8090", fetch: stubFetch.fetch });
  const media = new Media(transport);

  // await silo.media.upload({ bytes, filename: "hero.png", contentType: "image/png", folder: "heroes" })
  const bytes = new Uint8Array([1, 2, 3]);
  stubFetch.enqueue(StubResponse.json(assetPayload({ id: "01UPLOAD1" }), 201));
  await media.upload({ bytes, filename: "hero.png", contentType: "image/png", folder: "heroes" });

  // await silo.media.upload(file, { folder: "heroes" })     // a browser File, or a Blob
  const file = new File(["bytes"], "photo.jpg", { type: "image/jpeg" });
  stubFetch.enqueue(StubResponse.json(assetPayload({ id: "01UPLOAD2" }), 201));
  await media.upload(file, { folder: "heroes" });

  // const library = await silo.media.list({ ... })
  stubFetch.enqueue(
    StubResponse.json(
      { items: [assetPayload({ id: "a" }), assetPayload({ id: "b" })], total: 3, limit: 50, offset: 0 },
      200,
    ),
  );
  const library = await media.list({
    text: "hero",
    folder: "heroes",
    recursive: true,
    extension: "png",
    tag: "banner",
    modifiedAfter: "2026-01-01",
    modifiedBefore: "2026-09-01",
    limit: 50,
    offset: 0,
    sort: "-updated_at",
  });
  expect(library.files.map((asset) => asset.id)).toEqual(["a", "b"]);
  expect(library.total).toBe(3);

  stubFetch.enqueue(StubResponse.json({ items: [assetPayload({ id: "c" })], total: 3, limit: 50, offset: 50 }, 200));
  const nextPage = await library.next();
  expect(nextPage?.files.map((asset) => asset.id)).toEqual(["c"]);

  // for await (const asset of silo.media.all({ folder: "heroes" })) { /* ... */ }
  stubFetch.enqueue(StubResponse.json({ items: [assetPayload({ id: "only" })], total: 1, limit: 50, offset: 0 }, 200));
  const seen: string[] = [];
  for await (const asset of media.all({ folder: "heroes" })) seen.push(asset.id);
  expect(seen).toEqual(["only"]);

  // await silo.media.extensions()
  stubFetch.enqueue(StubResponse.json({ items: ["png", "jpg"] }));
  expect(await media.extensions()).toEqual(["png", "jpg"]);

  // const asset = await silo.media.get(id)
  stubFetch.enqueue(StubResponse.json(assetPayload()));
  const asset = await media.get("01ASSET");
  expect(asset.url).toBe("http://localhost:8090/media/01ASSET");
  expect(asset.reference).toBe("silo://media/01ASSET");
  expect(asset.filename).toBe("hero.png");
  expect(asset.folder).toBe("heroes");
  expect(asset.sizeInBytes).toBe(2048);
  expect(asset.contentType).toBe("image/png");
  expect(asset.tags).toEqual(["banner"]);
  expect(asset.usageCount).toBe(1);

  // await asset.rename("hero-2.png")
  stubFetch.enqueue(StubResponse.json(assetPayload({ filename: "hero-2.png" })));
  await asset.rename("hero-2.png");
  expect(asset.filename).toBe("hero-2.png");

  // await asset.moveTo("heroes/2026")
  stubFetch.enqueue(StubResponse.json(assetPayload({ filename: "hero-2.png", folder: "heroes/2026" })));
  await asset.moveTo("heroes/2026");
  expect(asset.folder).toBe("heroes/2026");

  // await asset.setTags(["banner", "hero"])   // replaces the list, which is what PATCH does
  stubFetch.enqueue(StubResponse.json(assetPayload({ tags: ["banner", "hero"] })));
  await asset.setTags(["banner", "hero"]);
  expect(asset.tags).toEqual(["banner", "hero"]);

  // await asset.delete()                      // refused while an entry references it
  stubFetch.enqueue(
    StubResponse.errorBody(409, "media_in_use", "still referenced", {
      usage_count: 1,
      visible_count: 1,
      visible_capped: false,
      referrers: [],
    }),
  );
  await expect(asset.delete()).rejects.toBeInstanceOf(MediaInUseError);

  // await asset.delete({ force: true })       // also needs entries:update where it reaches
  stubFetch.enqueue(StubResponse.empty());
  await asset.delete({ force: true });

  // const usage = await asset.usages()
  stubFetch.enqueue(
    StubResponse.json({
      items: [{ media_id: "01ASSET", project: "acme", env: "prod", collection: "posts", entry_id: "01K1" }],
      total: 4,
      visible: 1,
      visible_capped: true,
    }),
  );
  const usage = await asset.usages();
  expect(usage.usages).toEqual([
    { mediaId: "01ASSET", project: "acme", environment: "prod", collection: "posts", entryId: "01K1" },
  ]);
  expect(usage.total).toBe(4);
  expect(usage.visible).toBe(1);
  expect(usage.visibleCapped).toBe(true);

  // MediaReference.idOf("silo://media/01J8...")     // "01J8..." | null
  expect(MediaReference.idOf(MediaReference.of(asset.id))).toBe(asset.id);

  // const report = await silo.media.deleteMany(ids, { force: true })
  stubFetch.enqueue(
    StubResponse.json({ deleted: ["a", "b"], failed: [{ id: "c", code: "not_found", message: "gone" }] }),
  );
  const report = await media.deleteMany(["a", "b", "c"], { force: true });
  expect(report.deleted).toEqual(["a", "b"]);
  expect(report.failed).toEqual([{ id: "c", code: "not_found", message: "gone" }]);

  // await silo.media.folders.list()
  stubFetch.enqueue(StubResponse.json({ items: ["heroes"] }));
  expect(await media.folders.list()).toEqual(["heroes"]);

  // await silo.media.folders.create("heroes/2026")
  stubFetch.enqueue(StubResponse.json({ path: "heroes/2026" }, 201));
  expect(await media.folders.create("heroes/2026")).toBe("heroes/2026");

  // await silo.media.folders.rename("heroes", "banners", { merge: true })
  stubFetch.enqueue(StubResponse.json({ from: "heroes", to: "banners" }));
  expect(await media.folders.rename("heroes", "banners", { merge: true })).toEqual({ from: "heroes", to: "banners" });

  // await silo.media.folders.delete("banners", { recursive: true, force: true })
  stubFetch.enqueue(StubResponse.empty());
  await media.folders.delete("banners", { recursive: true, force: true });
});

File: packages/silo-client-java/src/test/java/in/org/quicko/silo/client/MediaExamplesTest.java
package in.org.quicko.silo.client;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import in.org.quicko.silo.client.errors.MediaInUseException;
import in.org.quicko.silo.client.media.MediaAsset;
import in.org.quicko.silo.client.media.MediaDeleteReport;
import in.org.quicko.silo.client.media.MediaFolderDeleteOptions;
import in.org.quicko.silo.client.media.MediaFolderMoveOptions;
import in.org.quicko.silo.client.media.MediaFolders;
import in.org.quicko.silo.client.media.MediaPage;
import in.org.quicko.silo.client.media.MediaQuery;
import in.org.quicko.silo.client.media.MediaReplace;
import in.org.quicko.silo.client.media.MediaUpload;
import in.org.quicko.silo.client.media.MediaUsagePage;
import in.org.quicko.silo.client.support.StubHttp;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

/** The README's media examples, run. Its siblings are in {@link ExamplesTest}. */
class MediaExamplesTest {
  private static final String Asset =
      "{\"id\":\"m1\",\"filename\":\"logo.png\",\"folder\":\"brand\",\"blob_key\":\"ab/cd.png\","
          + "\"size\":2048,\"content_type\":\"image/png\",\"hash\":\"deadbeef\",\"state\":\"active\","
          + "\"tags\":[\"brand\"],\"url\":\"https://silo.example.com/media/m1\","
          + "\"created_at\":\"2026-09-01T10:00:00.000Z\",\"updated_at\":\"2026-09-01T10:00:00.000Z\","
          + "\"usage_count\":2}";

  private StubHttp server;

  @TempDir
  Path workspace;

  @BeforeEach
  void newStub() {
    server = StubHttp.create();
  }

  private Path file(String name) throws IOException {
    Path path = workspace.resolve(name);
    Files.write(path, "bytes".getBytes(StandardCharsets.UTF_8));
    return path;
  }

  @Test
  void anEntryStoresTheReferenceAndNeverTheUrl() throws IOException {
    server.enqueueJson(Asset);

    MediaAsset logo = server.silo("k").media().upload(
        MediaUpload.of(file("logo.png")).folder("brand"));

    assertEquals("silo://media/m1", logo.reference());
    assertTrue(logo.url().endsWith("/media/m1"));
    assertTrue(server.lastRequest().body().contains("filename=\"logo.png\""));
    assertTrue(server.lastRequest().body().contains("name=\"folder\""));
  }

  @Test
  void aMutatingCallAdoptsTheServersAnswerInPlace() {
    server.enqueueJson(Asset);
    server.enqueueJson(Asset.replace("logo.png", "logo-2026.png"));
    server.enqueueJson(Asset.replace("\"folder\":\"brand\"", "\"folder\":\"brand/archive\""));
    server.enqueueJson(Asset.replace("[\"brand\"]", "[\"brand\",\"archived\"]"));

    MediaAsset asset = server.silo("k").media().get("m1");

    asset.rename("logo-2026.png");
    assertEquals("logo-2026.png", asset.filename());

    asset.moveTo("brand/archive");
    assertEquals("brand/archive", asset.folder());

    asset.setTags(List.of("brand", "archived"));
    assertEquals(List.of("brand", "archived"), asset.tags());
  }

  @Test
  void replacingTheBytesKeepsTheReferenceAndMovesTheHash() throws IOException {
    server.enqueueJson(Asset);
    server.enqueueJson(Asset
        .replace("\"size\":2048", "\"size\":4096")
        .replace("\"hash\":\"deadbeef\"", "\"hash\":\"cafebabe\""));

    MediaAsset asset = server.silo("k").media().get("m1");
    asset.replace(MediaReplace.of(file("logo-v2.png")));

    assertEquals("cafebabe", asset.hash());
    assertEquals(4096, asset.sizeInBytes());
    assertEquals("silo://media/m1", asset.reference(), "the reference survives a replace");
    assertEquals("/api/media/m1/content", server.lastRequest().path());
  }

  @Test
  void listingTakesAQueryAndPagesLazily() {
    server.enqueueJson("{\"items\":[" + Asset + "],\"total\":1,\"limit\":50,\"offset\":0}");
    server.enqueueJson("{\"items\":[" + Asset + "],\"total\":1,\"limit\":50,\"offset\":0}");
    server.enqueueJson("{\"items\":[\"png\",\"svg\"]}");

    MediaPage page = server.silo("k").media().list(MediaQuery.all()
        .folder("brand")
        .recursive(true)
        .extension("png")
        .modifiedAfter(Instant.now().minus(Duration.ofDays(30))));
    assertEquals(1, page.files().size());

    String target = server.request(0).target();
    assertTrue(target.contains("folder=brand"), target);
    assertTrue(target.contains("recursive=true"), target);
    assertTrue(target.contains("ext=png"), target);
    assertTrue(target.contains("modified_after="), target);

    List<String> filenames = new ArrayList<>();
    for (MediaAsset file : server.silo("k").media().all(MediaQuery.all().text("logo"))) {
      filenames.add(file.filename());
    }
    assertEquals(List.of("logo.png"), filenames);

    assertEquals(List.of("png", "svg"), server.silo("k").media().extensions());
  }

  @Test
  void foldersAreExplicitRecords() {
    server.enqueueJson("{\"path\":\"brand/2026\"}");
    server.enqueueJson("{\"from\":\"brand/2026\",\"to\":\"brand/current\"}");
    server.enqueueNoContent();

    MediaFolders folders = server.silo("k").media().folders();

    assertEquals("brand/2026", folders.create("brand/2026"));
    assertEquals("brand/current",
        folders.rename("brand/2026", "brand/current", MediaFolderMoveOptions.merging()).to());
    assertTrue(server.request(1).body().contains("\"merge\":true"));

    folders.delete("brand/archive", MediaFolderDeleteOptions.recursively().force(true));
    String target = server.lastRequest().target();
    assertTrue(target.contains("path=brand%2Farchive"), target);
    assertTrue(target.contains("recursive=true"), target);
    assertTrue(target.contains("force=true"), target);
  }

  @Test
  void aRefusedDeleteCarriesTheFactsNeededToDecide() {
    server.enqueueJson(Asset);
    server.enqueue(409, "application/json",
        "{\"error\":{\"code\":\"media_in_use\",\"message\":\"still referenced\",\"details\":"
            + "{\"usage_count\":3,\"visible_count\":2,\"visible_capped\":true,\"referrers\":"
            + "[{\"media_id\":\"m1\",\"project\":\"acme\",\"env\":\"prod\","
            + "\"collection\":\"posts\",\"entry_id\":\"e1\"}]}}}");

    MediaAsset asset = server.silo("k").media().get("m1");

    MediaInUseException caught =
        assertThrows(MediaInUseException.class, () -> asset.delete());

    assertEquals(3, caught.usageCount());
    assertEquals(2, caught.visibleCount());
    assertTrue(caught.visibleCapped());
    assertEquals("posts", caught.referrers().get(0).collection());
  }

  @Test
  void usagesPageByWhatThisKeyMaySee() {
    server.enqueueJson(Asset);
    server.enqueueJson("{\"items\":[{\"media_id\":\"m1\",\"project\":\"acme\",\"env\":\"prod\","
        + "\"collection\":\"posts\",\"entry_id\":\"e1\"}],\"total\":9,\"visible\":1,"
        + "\"visible_capped\":false}");

    MediaUsagePage usages = server.silo("k").media().get("m1").usages();

    assertEquals(9, usages.total(), "the true referrer count");
    assertEquals(1, usages.visible(), "what this key is allowed to see");
  }

  @Test
  void aBulkDeleteAnswersAReportRatherThanThrowing() {
    server.enqueueJson("{\"deleted\":[\"m1\"],\"failed\":[{\"id\":\"m2\","
        + "\"code\":\"media_in_use\",\"message\":\"still referenced\",\"usage_count\":2,"
        + "\"visible_count\":2,\"visible_capped\":false,\"referrers\":[]}]}");

    MediaDeleteReport report = server.silo("k").media().deleteMany(List.of("m1", "m2"));

    assertEquals(List.of("m1"), report.deleted());
    assertEquals("m2", report.failed().get(0).id());
    assertEquals("media_in_use", report.failed().get(0).code());
  }
}
